server/src/com/sixsq/slipstream/ssclj/resources/spec/identification.cljc

```diff
@@ -0,0 +1,29 @@
+(ns com.sixsq.slipstream.ssclj.resources.spec.identification
+  (:require
+    [clojure.spec.alpha :as s]
+    [com.sixsq.slipstream.ssclj.resources.spec.common :as cimi-common]
+    [com.sixsq.slipstream.ssclj.resources.spec.core :as cimi-core]
+    [com.sixsq.slipstream.ssclj.util.spec :as su]))
+
+; {
+;   "user": string,
+; 	"idkey": string,
+; 	"deviceid": string
+; }
+
+
+(s/def :cimi.identification/user ::cimi-core/nonblank-string)
```

Comment:
I guess this is for linking the "field" to another resource, but I dont understand how to establish such link... in any case, is it problematic if I store directly the user here?

Reply:
look at the `device-dynamic` resource from Souvik. There's a link to `device`

Reply:
I have removed the "user" field, I decided that I don't need it and thus, there is nothing to link

Reply:
but you do need it, otherwise it can't be associated with a user, leaving room for impersonation attacks

Reply:
It can... in order to register a new device, the user credentials will be required. It means I will have the validated username at hand

Reply:
> in order to register a new device

we are not talking about devices. Picture the following scenario:
 - I sign up for mf2c in the dashboard
    - a new user is created for me in CIMI, plus, an IDKey is generated and also written in CIMI, by mF2C, resulting in the following resource: `{id: <uuid>, idkey: <generatedByIdentification>, created: <timestamp>, updated: <timestamp>, acl: { owner: ADMIN, rules: [ADMIN] } }`
 - then I download the compose file. There's the respective IDKey embedded in there, as an env variable
 - I install that compose file in my RPi
    - all goes well. The CAU will take that IDkey and double check with the cloud that IDKey exists. And if fact it does, so the agent is considered as legit
 - now someone looks over my shoulder and steals my IDKey, pastes in their own compose file, and install it in their own malicious computer 
     - all goes well too, because the CAU will again take that IDKey, verify that it exists, and because it does, it considers it as legit.

So you see, that's not good. You need a user reference, so that whenever the CAU does something like: `GET /api/identification?$filter=idkey=<idkey>`, CIMI will take actions and notify the respective user by email, that a new device has been added. And we cannot do this unless we know which user the idkey belongs to.

Reply:
Ok, but that scenario won't be valid in the near future. After this resource is published and I can use it, the idkey wont be in the docker-compose file anymore...  Instead of that, the user credentials must to be provided and the idkey will be retrieved from the cloud. Dont forget that that is the purpose of this resource (see issues 192 and 152).

Reply:
let's go by steps:
>that scenario won't be valid in the near future
if I am a reviewer, this would be my first scenario to challenge the system security

>the idkey wont be in the docker-compose file anymore 
it will. That's what's described in D3.4 which was just submitted. If you're referring to figure 7 from D3.4, there are many things unexplained and insecure in that workflow

> Instead of that, the user credentials must to be provided
how? who provides those credentials? this is quite a disruptive change. Something which is not in any deliverable nor planned for mf2c, given the amount of time we have left. 
you're free to implement this, if you have a proper way to automate the agent deployment, without the need for the user to interactively input any credentials, but for mf2c, D3.4 and ticket 152, the IDKey is present in the compose file, and it is the CAU who will verify its authenticity

I don't see any issue 192. As for #152 (https://github.com/mF2C/mF2C/issues/152), the purpose is driven by figures 22 and 8, and it assumes IDKey in the compose file, as well as the same IDKey registering in the cloud agent through CIMI